src/hooks/useQuery.ts

```diff
@@ -28,61 +29,54 @@ export type UseQueryResponse<T> = [
 export const useQuery = <T = any, V = object>(
   args: UseQueryArgs<V>
 ): UseQueryResponse<T> => {
-  const isMounted = useRef(true);
   const unsubscribe = useRef(noop);
   const client = useContext(Context);
 
-  // This creates a request which will keep a stable reference
-  // if request.key doesn't change
-  const request = useRequest(args.query, args.variables);
-
-  const [state, setState] = useState<UseQueryState<T>>({
+  // This is like useState but updates the state object
+  // immediately, when we're still before the initial mount
+  const [state, setState] = useImmediateState<UseQueryState<T>>({
     fetching: false,
-    error: undefined,
     data: undefined,
+    error: undefined,
   });
 
+  // This creates a request which will keep a stable reference
+  // if request.key doesn't change
+  const request = useRequest(args.query, args.variables);
+
   const executeQuery = useCallback(
     (opts?: Partial<OperationContext>) => {
       unsubscribe.current();
 
+      // If useQuery is currently paused, set fetching to
+      // false and abort; otherwise start the query
       if (args.pause) {
         setState(s => ({ ...s, fetching: false }));
         unsubscribe.current = noop;
         return;
+      } else {
+        setState(s => ({ ...s, fetching: true }));
       }
 
-      setState(s => ({ ...s, fetching: true }));
-
-      const [teardown] = pipe(
+      [unsubscribe.current] = pipe(
```

Comment:
A bit confused by how this works. Is this array destructuring also capable of reassignment, i.e. this basically turns into:

```
const [teardown] = pipe(/* stuff */)
unsubscribe.current = teardown;
```

It seems like a dope feature of TS if true.

Reply:
Yep, works in pure JS as well, since the destructuring target can almost be any expression afaik